juju/client/connection.py

```diff
@@ -536,7 +537,19 @@ async def _do_ping():
             log.debug("ping failed because of closed connection")
             pass
 
-    async def rpc(self, msg: dict, encoder=None) -> dict:
+    @overload
+    async def rpc(
+        self, msg: dict[str, _JSON], encoder: None = None
+    ) -> dict[str, _JSON]: ...
+
+    @overload
+    async def rpc(
+        self, msg: dict[str, _RICH_JSON], encoder: TypeEncoder
+    ) -> dict[str, _JSON]: ...
+
+    async def rpc(
+        self, msg: dict[str, Any], encoder: json.JSONEncoder | None = None
+    ) -> dict[str, Any]:
```

Comment:
If it's a pain to narrow this definition, let's just leave it with the Any's, but just pointing it out in case you meant to narrow it further